atlasdb-client/src/main/java/com/palantir/atlasdb/util/TaggedMetricsInvocationEventHandler.java

```diff
@@ -73,16 +87,21 @@ public void onSuccess(@Nullable InvocationContext context, @Nullable Object resu
         }
 
         long nanos = System.nanoTime() - context.getStartTimeNanos();
+        TimerCacheKey cacheKey = ImmutableTimerCacheKey.of(context.getMethod(), tagFunction.apply(context));
+        timerCache.computeIfAbsent(cacheKey, this::registerTimer).update(nanos, TimeUnit.NANOSECONDS);
+    }
+
+    private Timer registerTimer(TimerCacheKey timerCacheKey) {
         MetricName finalMetricName = MetricName.builder()
-                .safeName(MetricRegistry.name(serviceName, context.getMethod().getName()))
-                .putAllSafeTags(tagFunction.apply(context))
+                .safeName(MetricRegistry.name(serviceName, timerCacheKey.method().getName()))
+                .putAllSafeTags(timerCacheKey.tags())
                 .build();
-        taggedMetricRegistry.timer(finalMetricName).update(nanos, TimeUnit.NANOSECONDS);
+        return taggedMetricRegistry.timer(finalMetricName);
```

Comment:
This is a lot of complexity, how confident are we that it performs better than the string concatenation approach?

What if we replace
```java
.safeName(MetricRegistry.name(serviceName, context.getMethod().getName()))
```
with
```java
.safeName(serviceName + '.' + context.getMethod().getName())
```

I imagine that will perform better, especially on java 9+ with [JEP 280](https://openjdk.java.net/jeps/280).